### ISSUE TITLE
Invalid JSON RPC response: {"id":15,"jsonrpc":"2.0","error":{"code":-32603}}

### ISSUE THREAD

**Issue by vpomo** [MetaMask/metamask-extension#3140]
Does not work from January 26, 2018. 
I, too, have such a mistake. She appeared since January 26, 2018. I know for sure that it was not on January 25, 2018. I implemented a smart contract without the error. On January 27, 2018, I was no longer able to make this smart contract a reality. The error manifests itself when a smart contract is made with parameters. On the main network, this error exists. On the network Ropsten it is not.
Everyone talks about this:
https://gitter.im/ethereum/remix

**Comment by vpomo** [comment]
It's still:
https://ethereum.stackexchange.com/questions/37806/invalid-json-rpc-response-when-creating-contract/37818#37818

The contract must be with parameters and nested other contracts

**Comment by citrixrep** [comment]
this problem fix using Brave, with firefox and chrome dont work on mainnet

**Comment by danfinlay** [comment]
We believe we found the source of this and that it has been fixed. Let us know if it is not fixed, and I will reopen the issue.

We recently switched our Infura connection to a RESTful API, to reduce network load on them. Part of that required allowing differently structured requests, to account for contract publishing payloads under the new API.

**Comment by aphexyuri** [comment]
I'm still experiencing this right now

**Comment by raoneel** [comment]
Still experiencing this as well on Remix. I can deploy some contracts, but for a larger contract I'm getting this exact error.

Network: Rinkeby
Browser: Chrome 63, Mac OSX

**Comment by MCKapur** [comment]
Having the same issue! :/

**Comment by allurco** [comment]
Please Re-open the issue, this is not fixed yet.

**Comment by olgierds** [comment]
@danfinlay 
Hi, it's still not fixed.

**Comment by danfinlay** [comment]
Can anyone please provide reprosuction steps, ideally for a test network? We have not reproduced this problem.

**Comment by danfinlay** [comment]
Ok we may have a reproduction contract here:
https://github.com/MetaMask/metamask-extension/issues/3117#issuecomment-362625939

**Comment by sameepsi** [comment]
Yes I am still able to reproduce it using remix.
Here are the steps-:
1. The sample code is available in the following gist- https://gist.github.com/sameepsi/c2e09cc0782e6dbf0ae0980a90e9098f
2. The contract parameters are 10000000, "TEST", "test"

**Comment by danfinlay** [comment]
Alright, thanks @sameepsi, I've been able to reproduce this issue, Infura is throwing a `500` error in response to this constructor. I suspect it has to do with some changes we recently made to our API connection to them.

We should have this fixed soon, but in the meanwhile, you should be able to work around it by clicking the top-left provider menu in MetaMask, and entering `https://mainnet.infura.io/500workaround`, or the address of another ethereum client running an RPC service.

**Comment by olgierds** [comment]
Thanks,
rinkeby also works on
https://rinkeby.infura.io/500workaround

**Comment by massimo110** [comment]
same error in https://rinkeby.infura.io/500workaround
Invalid JSON RPC response: {"id":824,"jsonrpc":"2.0","error":{"code":-32601}}

**Comment by massimo110** [comment]
https://rinkeby.infura.io/500workaround   not work for me.

**Comment by chebykin** [comment]
In my **private** aura network, I'm facing the same error.
I use Remix <-> MetaMask v3.14.1 (Chrome) <-> Parity v1.9.2

- transactions with state modification - fail with 32603 error
- transactions without state modifications - calls work fine

**Comment by danfinlay** [comment]
This should be fixed. For those reporting this is not fixed (@olgierds, @massimo110, @chebykin), please provide fresh reproduction steps, since the excellent steps [provided above](https://github.com/MetaMask/metamask-extension/issues/3140#issuecomment-362631772) no longer cause a problem.

**Comment by oschvr** [comment]
Im still getting the error at remix when I click debug, I've tried all steps [above](https://github.com/MetaMask/metamask-extension/issues/3140#issuecomment-362631772) and also tried to use the [/500workaround](https://github.com/MetaMask/metamask-extension/issues/3140#issuecomment-362652301)

**Comment by danfinlay** [comment]
The `/500workaround` endpoint is now irrelevant, its fix should be deployed to all networks.

Since no core devs can reproduce, the debugging needs to be done by someone who can reproduce.

Try opening the [background inspector](http://metamask.helpscoutdocs.com/article/21-how-to-get-logs-and-help-metamask-support-and-diagnose-your-issue), and looking at the `Network` tab when making this request. Click on the request that should be publishing, and look at its response. What's the error?

Bonus: Right click that request, click `copy as cURL`, and send it to us, maybe at support@metamask.io, so we can debug very precisely at the provider layer.

**Comment by chebykin** [comment]
@danfinlay, thank you for the instructions!

As for me, it looks like a problem with Parity aura engine or it's config. This is a request/response pair from the Parity node rpc logs:

```2018-02-14 15:17:10   TRACE rpc  Request: {"id":2637171441,"jsonrpc":"2.0","params":[{"from":"0x3c6463e6bb1d5879c54a2240545e3a629cf09358","to":"0xaa488f54fa157e8dfee11561ea40b2ba6ef161ba","value":"0x0","data":"0xbc201a9d0000000000000000000000000daa3d4b5a83ec4175277027483657165452a242"}],"method":"eth_estimateGas"}.```

```2018-02-14 15:17:10   DEBUG rpc  Response: {"jsonrpc":"2.0","error":{"code":-32015,"message":"Transaction execution error.","data":"Internal(\"Requires higher than upper limit of 47000000\")"},"id":2637171441}.```

The two frustrating things about MetaMask response are:

- a different than in the server response error code
- a dropped off error description.

**Comment by sameepsi** [comment]
@chebykin As far as I can see from the logs you need to increase the gas limit for the deployment of your contract.

**Comment by danfinlay** [comment]
> Will try to find something in Parity issue tracker.

Let us know how that goes.

I'm a little curious why some people reported this isn't working on Infura, though. They are running Geth right now, not Parity.

**Comment by chebykin** [comment]
In my case, it was my mistake with chain state which lead to the response I described above. I tried to invoke gas estimation request against the brand new chain with no blocks mined yet. The error response disappeared after the first block was mined.

As I see here https://github.com/kumavis/json-rpc-engine/blob/b5185e5d6e75717745269f5f4be65af42bf6bb40/src/index.js#L107 the -32602 error is a default one, so it could wrap any error from the server.

**Comment by danfinlay** [comment]
This bug sounds possibly related to the one described in #3285. For some reason many people are getting out of gas errors when estimating gas, both on geth and parity.

Closing this issue in favor of that one, because this issue dealt with a different problem that was fixed, and that one includes much more specific investigation into that issue.

**Comment by jtakalai** [comment]
Getting that error when trying to open debugger. In Remix, open debugger tab, enter block number and tx, press Play. Get ```Invalid JSON RPC response: {"id":1763,"jsonrpc":"2.0","error":{"code":-32603}}```

**Comment by jessicagmarshall** [comment]
@kumavis I managed to fix this by updating truffle to the most recent beta version (Truffle v4.1.0 (core: 4.1.0)). I take back what I said about thinking this was a MetaMask issue.

**Comment by antogh** [comment]
I'm having the same issue described by @jtakalai above.

I use remix + metamask inside chrome, on the rinkeby testnet.

All work fine when I deploy a contract or interact with previously deployed contract at some address.

The failure comes when I use the debugger tab to debug a certain transaction. I enter the block/tx paramaeters , then click  start debugging and, boom, I get the error:
"Invalid JSON RPC response: {"id":30,"jsonrpc":"2.0","error":{"code":-32603}}"

Notice that if in Remix I change the environment from the metamask provided "injected web3" to my local geth node "web3 provider" the debugger with the very same transaction works flawlessly, so the problem must be in the way metamask is handling the bridge for this kind of operation. 

Any help, workaround to solve this issue greatly appreciated, keeping a local node always in sync with Eth blockchain is already annoying on rinkeby, don't want to imagine how hard can be on the main net. 

Thank you

**Comment by danfinlay** [comment]
You can work around this for now by entering this custom RPC endpoint:
https://mainnet.infura.io/metamask

Where”mainnet” can be most network names.

**Comment by antogh** [comment]
thanks Dan,
where i should enter this setting exactly? Remix settings, metamask?
and ...
should I use
https://rinkeby.infura.io/metamask 
since I am on the rinkeby net?

**Comment by antogh** [comment]
just tried 
https://rinkeby.infura.io/metamask
in remix as web3 provider endpoint
interacts correctly with a deployed contract 
but starting debugger on the prevsiously mentioned tx produces always 
Invalid JSON RPC response: ""
so to be able to debug I'm stuck with my geth local node that needs to download all the ethereum blockchain

**Comment by danfinlay** [comment]
You should enter the custom rpc url in MetaMask. Find “Custom RPC” in the network menu when you click the network name in MetaMask.

**Comment by antogh** [comment]
Set the custom RPC as you suggested, used address 
https://rinkeby.infura.io/metamask
now MetaMask says "Private network" on top left corner
tried remix: interacting with the contract , reading functions works fine, but...
trying to start debugger on any transaction gives me the same error :((( with slight different code:
Invalid JSON RPC response: {"id":68,"jsonrpc":"2.0","error":{"code":-32601}}
before was 
Invalid JSON RPC response: {"id":30,"jsonrpc":"2.0","error":{"code":-32603}}
any new idea?

**Comment by danfinlay** [comment]
Oh, MetaMask doesn't support the remix debugger yet. Sorry I didn't understand that's what you were trying to do. We do have a PR for adding it in client-side, and it may be merged soon, but it does not work yet.

We're also going to be fixing these undescriptive errors soon.

**Comment by antogh** [comment]
What a pity :( ...
This would be a killer feature, AFAIK there is no way to debug a TX with source trace without having a local node in sync with the whole ETH blockchain, this is time and resource consuming. Etherscan has a debug feature but without the source trace (useless IMO)
If you provide such a tool it would be a relief for many. 
If you look at the  @jtakalai  post above for example he is trying to do the same I guess.
Hope to receive good news from Metamask soon.

**Comment by danfinlay** [comment]
Yes, it would be a killer feature...
https://github.com/MetaMask/metamask-extension/pull/3482

**Comment by fan31415** [comment]
I get the same error too, luckly I fix it later. Hope these info will help :)
#### Environment
ganache-cli in docker
truffle:4.1.3
soldity:0.4.19
#### Reproduction
1. I add a function in my contract
2. use `truffle migration` to deploy my contract
3. invoke this function in js by using truffle-contract's contract instance
4. encounter this error.
#### Solution
Use `truffle migration --reset` in step 2.

**Comment by ltfschoen** [comment]
I fixed this error by updating to the latest Node.js, Truffle, Ganache-CLI, i.e.
```
$ nvm use v9.8.0;
$ node --version
v9.8.0
$ npm install -g truffle ganache-cli;
$ truffle --version
Truffle v4.1.3 - a development framework for Ethereum
$ ganache-cli version
Ganache CLI v6.1.0 (ganache-core: 2.1.0)
``` 
I then restarted Ganache CLI and restarted my front-end, and then my code worked without the error

**Comment by jetaggart** [comment]
I received this error using firefox and a sample app

**Environment**
Firefox 59.0.2 (64-bit)
macOS 10.13.13 
Metamask 3.13.8
https://github.com/f-o-a-m/purescript-web3-example -- built from scratch no modifications

** Reproduction **
1. Follow steps in repo to build sample app
2. Select Metamas from dropdown
3. Navigate to http://0.0.0.0:8080/ in Firefox
4. Console error: Error: Invalid JSON RPC response: "", Error Code: -32603

**Solution**
Use chrome

**Comment by ltfschoen** [comment]
It was also important for me to redeploy the contracts if I restarted TestRPC with `truffle migrate --reset --network development;`

**Comment by bitdom0** [comment]
Same issue using Firefox 59.0.2 and MetaMask  3.13.8.

**Comment by danfinlay** [comment]
> I received this error using firefox and a sample app. The application works fine in Chrome 

Firefox has recently been a bit slower about deploying our latest fixes, we are working with them to improve this process. It's very possible this bug has been fixed in our latest release, but on Firefox we need to get our builds perfectly deterministic before we can get reliable deploys anymore.

**Comment by showmeyourcode** [comment]
**Environment**
Chrome : Version 65.0.3325.181 (Official Build) (64-bit)
GANACHE : Version 1.1.0 (1.1.0)
Metamask : 4.5.2 (Had to donwgrade because of [this](https://github.com/MetaMask/metamask-extension/issues/3881) issue)

**web3.js code**
```
  //Check Game State
  game.state.contract.methods.getGameState().call((err, result) => {
      if (err) {
        console.error(err);
        return;
      }
  }

```

**RRROROUTPUT**
```
Error: Returned error: {"id":2,"jsonrpc":"2.0","error":{"code":-32603}}


```

**SOLIDITY CODE**
```js
function getGameState() public view returns(uint256 _gameNumber, bool _registerationOpen, bool _playStarted, bool _revealing, bool _lastGameFinished, uint256 _totalPlayers) {
        return (games.length, registerationOpen, playStarted, revealing, lastGameFinished, games[games.length-1].players.length);
}
    
```

**Comment by j10sanders** [comment]
I'm getting this issue from chrome, and I am using truffle develop.

Truffle v4.1.5 (core: 4.1.5)
MetaMask 4.5.5

**Comment by dmi3y** [comment]
In my case all I had to do is to connect account (with funds) to metamask

**Comment by bdresser** [comment]
@bitpshr was this fixed by #4433 ?

**Comment by bitpshr** [comment]
@bdresser I don't think this is related to #4433. That PR fixed a specific issue where certain RPC errors were only being logged on the client, not actually thrown. This sounds like a different FireFox-specific issue and one that may be fixed already, it just needs investigation.

**Comment by bdresser** [comment]
@psylinse @bitdom0 firefox is now up to date with Chrome & our most recent release (4.7.4) – are you still seeing this issue?

**Comment by xenu256** [comment]
At least for me, all rpc 32603 errors are due to my own fault, like 3 reasons: 1) not having latest contract json in dApp, 2) migrations not reset, or 3) wrong network (json specifies one, you try to connect to another). Metamask works fine.

**Comment by birdkingz** [comment]
I'm having 
`MetaMask - RPC Error: Internal JSON-RPC error. {code: -32603, message: "Internal JSON-RPC error."}`
while calling `estimateGas` on my contract.

The contract is working fine because if I don't use metamask then I'll be able to call the function.

**Comment by danfinlay** [comment]
@birdkingz This thread has become a catch-all for people receiving this generic error code. Please open a new issue, and include specific reproduction steps, such as a Dapp or your contract's address and the transaction whose estimation fails.

**Comment by kumavis** [comment]
**If you encounter an issue similar to this, please open a new issue, and include specific reproduction steps, such as a Dapp or your contract's address and the transaction whose estimation fails.**

**Comment by m1z4rd** [comment]
Metamask Team, your wallet is so rad.

But there's still an error deploying on Remix Plugin for vscode.

**there are errors deploying: error: invalid json rpc response: {"id":36,"jsonrpc":"2.0"}**
I had to use ganache instead.

Hopefully, you can fix it.


Thanks in advance

**Comment by kumavis** [comment]
@m1z4rd filed a new issue here https://github.com/MetaMask/metamask-extension/issues/12788
please add more information about how you encountered this
